File: scripts/ops/evidence.py
"""Pure reconciliation. Unknown is preferable to joining on a short SHA or time."""
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
import re
from urllib.parse import quote, urlparse

from .sources import full_sha


def utc(value) -> str:
    try:
        if isinstance(value, (int, float)):
            date = datetime.fromtimestamp(value / 1000, timezone.utc)
        elif value:
            date = datetime.fromisoformat(value.replace("Z", "+00:00"))
            date = date.replace(tzinfo=timezone.utc) if date.tzinfo is None else date.astimezone(timezone.utc)
        else:
            return ""
        return date.isoformat(timespec="milliseconds").replace("+00:00", "Z")
    except (ValueError, TypeError, OverflowError, OSError):
        return ""


def safe_url(value) -> str:
    if not isinstance(value, str):
        return ""
    try:
        parsed = urlparse(value)
        if parsed.scheme != "https" or not parsed.hostname or parsed.username or parsed.password:
            return ""
        # No query strings: signed URLs and access tokens must never reach output.
        return f"https://{parsed.netloc}{parsed.path}".rstrip("/")
    except ValueError:
        return ""


@dataclass
class Event:
    id: str
    kind: str
    app: str | None
    sha: str | None
    environment: str | None
    time: str
    state: str
    source: str
    url: str = ""
    updated_at: str = ""
    context_sha: str | None = None
    name: str = ""
    input_hash: str | None = None
    expires_at: str = ""
    current: bool | None = None
    notes: list[str] = field(default_factory=list)
    related: list[str] = field(default_factory=list)


def newest(events):
    return sorted(events, key=lambda e: (e.time, e.id), reverse=True)


def reconcile(snapshot: dict, apps: list[str]) -> list[Event]:
    repo = snapshot["repo"]
    base = f"https://github.com/{repo}"
    tags = snapshot["tags"]
    runs = {r["id"]: r for r in snapshot["runs"]}
    jobs = {j["id"]: j for j in snapshot["jobs"]}
    run_targets = {}
    run_envs = {}
    for run_id, run in runs.items():
        path = run.get("path", "").split("/")[-1]
        run_tags = [t for t in tags if t["run_id"] == run_id and t["name"].startswith("deploy/")]
        targets = {t["sha"] for t in run_tags}
        envs = {t["name"].split("/")[1] for t in run_tags}
        if (path == "cd-staging.yml" or path.startswith("ci-")) and run.get("event") in ("push", "workflow_dispatch"):
            targets.add(full_sha(run.get("head_sha")))
        run_targets[run_id] = next(iter(targets)) if len(targets) == 1 else None
        run_envs[run_id] = ("staging" if path == "cd-staging.yml" else "production" if path == "cd-production.yml"
                           else next(iter(envs)) if len(envs) == 1 else None)

    def job_app(job):
        match = re.fullmatch(r"(?:Build|Resolve|Deploy) (.+?)(?: Artifact)? \((?:Staging|Production|Rollback)\)", job.get("name", ""))
        app = match[1].lower().replace(" ", "-") if match else None
        return app if app in apps or app == "convex" else None

    events = []
    for run in runs.values():
        run_id = run["id"]
        context = full_sha(run.get("head_sha"))
        target = run_targets[run_id]
        note = [] if target else ["Requested checkout SHA unknown; workflow head is NOT the deployment target"]
        events.append(Event(f"run:{run_id}", "workflow", None, target, run_envs[run_id], utc(run.get("created_at")),
                            run.get("conclusion") or run.get("status", "unknown"), run.get("path", "").split("/")[-1],
                            f"{base}/actions/runs/{run_id}", utc(run.get("updated_at")), context,
                            name=f"attempt {run.get('run_attempt', 1)}", notes=note))
    for job in jobs.values():
        run_id = job["run_id"]
        # Reusable CI calls appear as "ci-web / Build & Bundle Size"; standalone
        # CI runs identify the app in the workflow filename. A successful Build
        # step is evidence of a CI build, not a prebuilt deployable upload.
        ci_name = re.fullmatch(r"ci-(.+) / Build(?: & Bundle Size)?", job.get("name", ""))
        path = runs.get(run_id, {}).get("path", "").split("/")[-1]
        ci_app = ci_name[1] if ci_name else path[3:-4] if path.startswith("ci-") and job.get("name") in ("Build", "Build & Bundle Size") else None
        if ci_app in apps:
            step = next((s for s in job.get("steps", []) if s.get("name") == "Build"), {})
            events.append(Event(f"ci-build:{job['id']}", "ci-build", ci_app, run_targets.get(run_id), None,
                                utc(step.get("started_at") or job.get("started_at")),
                                step.get("conclusion") or job.get("conclusion") or job.get("status", "unknown"),
                                f"run {run_id}", safe_url(job.get("html_url")),
                                utc(step.get("completed_at") or job.get("completed_at")),
                                full_sha(runs.get(run_id, {}).get("head_sha")), name=job.get("name", ""),
                                notes=["CI build only; no deployable bundle implied. PR checkout SHA not inferred from head_sha."]))
        app = job_app(job)
        if not app:
            continue
        deploy = job.get("name", "").startswith("Deploy ")
        events.append(Event(f"job:{job['id']}", "deploy-job" if deploy else "build-resolve", app,
                            run_targets.get(run_id), run_envs.get(run_id), utc(job.get("started_at")),
                            job.get("conclusion") or job.get("status", "unknown"), f"run {run_id}",
                            safe_url(job.get("html_url")), utc(job.get("completed_at")),
                            full_sha(runs.get(run_id, {}).get("head_sha")), name=job.get("name", ""),
                            notes=["Job result, not a serving-state or new-build guarantee"]))
    for artifact in snapshot["artifacts"]:
        app = next((a for a in sorted(apps, key=len, reverse=True) if artifact.get("name", "").startswith(a + "-")), None)
        if not app:
            continue
        suffix = artifact["name"][len(app) + 1:]
        match = re.fullmatch(r"(?:(?:staging|production|preview)-)?([0-9a-f]{16}|[0-9a-f]{40})", suffix)
        if not match:
            continue  # Test reports / CI build directories are not prebuilt CD bundles.
        run_id = (artifact.get("workflow_run") or {}).get("id")
        expiry = utc(artifact.get("expires_at"))
        expired = artifact.get("expired") is True or (expiry and expiry <= utc(snapshot["gathered_at"]))
        state = "expired" if expired else "available" if artifact.get("expired") is False and expiry else "unknown"
        sha = run_targets.get(run_id)
        events.append(Event(f"artifact:{artifact['id']}", "artifact", app, sha, None,
                            utc(artifact.get("created_at")), state, f"build run {run_id or 'unknown'}",
                            f"{base}/actions/runs/{run_id}/artifacts/{artifact['id']}" if run_id else "",
                            context_sha=full_sha((artifact.get("workflow_run") or {}).get("head_sha")),
                            name=artifact["name"], input_hash=match[1] if len(match[1]) == 16 else None, expires_at=expiry,
                            notes=["Upload time; checkout SHA from workflow evidence, not verified provenance",
                                   "Manifest, tarball checksum and SLSA attestation not downloaded or verified"] +
                                  ([] if sha else ["Builder checkout unknown; artifact head_sha is context only"])))
    for tag in tags:
        deploy = tag["name"].startswith("deploy/")
        events.append(Event("tag:" + tag["name"], "deploy-tag" if deploy else "release-tag", None,
                            tag["sha"], tag["name"].split("/")[1] if deploy else None,
                            utc(tag["time"]), "recorded" if deploy else "source-release", "git tag",
                            f"{base}/tree/{quote(tag['name'], safe='')}", name=tag["name"],
                            notes=["Repository-wide marker; does not prove every app deployed or is healthy"] if deploy
                            else ["Starter source version; not a deployable artifact or deployment"]))
    for row in snapshot["vercel"]:
        dep = row["deployment"]
        meta = dep.get("meta") or {}
        claims = {full_sha(meta.get(key)) for key in ("githubCommitSha", "gitCommitSha")}
        claims.discard(None)
        sha = next(iter(claims)) if len(claims) == 1 else None
        notes = ["Vercel git metadata; no artifact identity supplied"]
        if len(claims) > 1:
            notes.append("Conflicting commit metadata; SHA withheld")
        if meta.get("githubCommitOrg") and meta.get("githubCommitRepo") and f"{meta['githubCommitOrg']}/{meta['githubCommitRepo']}".lower() != repo.lower():
            sha = None
            notes.append("Metadata names a different repository; SHA withheld")
        events.append(Event(f"vercel:{dep['uid']}", "vercel", row["app"], sha, row["environment"],
                            utc(dep.get("createdAt") or dep.get("created")),
                            dep.get("readyState") or dep.get("state", "unknown"), "Vercel " + row["project"],
                            safe_url(dep["url"] if dep["url"].startswith("https://") else "https://" + dep["url"]) if dep.get("url") else "",
                            utc(dep.get("ready") or dep.get("readyAt")), current=row["current"], notes=notes))
    for dep in snapshot["deployments"]:
        statuses = sorted(dep.get("statuses", []), key=lambda s: (s.get("created_at", ""), s.get("id", 0)), reverse=True)
        status = statuses[0] if statuses else {}
        url = next((safe_url(s.get("environment_url")) for s in statuses if safe_url(s.get("environment_url"))), "")
        log = next((safe_url(s.get("log_url")) for s in statuses if safe_url(s.get("log_url"))), "")
        match = re.fullmatch(re.escape(base) + r"/actions/runs/(\d+)/job/(\d+)", log)
        run_id = int(match[1]) if match else None
        job = jobs.get(int(match[2]), {}) if match else {}
        app, target = job_app(job), run_targets.get(run_id)
        peers = [e for e in events if e.kind == "vercel" and url and e.url == url and e.environment == dep.get("environment") and (not app or e.app == app)]
        notes = ["GitHub environment status; inactive is not proof of Vercel serving state"]
        if len(peers) == 1:
            app = app or peers[0].app
            if target and peers[0].sha and target != peers[0].sha:
                notes.append("CONFLICT: workflow target and Vercel commit differ; SHA withheld")
                peers[0].notes.append("CONFLICT: GitHub workflow target differs; inspect both sources")
                target = None
            else:
                target = target or peers[0].sha
        if not target:
            notes.append("Deployment target unknown; GitHub deployment.sha is context only")
        event = Event(f"github-deployment:{dep['id']}", "github-deployment", app, target,
                      dep.get("environment"), utc(dep.get("created_at")), status.get("state", "unknown"),
                      f"GitHub deployment {dep['id']}", url or log, utc(status.get("created_at")),
                      full_sha(dep.get("sha")), notes=notes, related=[e.id for e in peers])
        for peer in peers:
            peer.related.append(event.id)
        events.append(event)
    # A commit/app join is a browsing link, NOT a claim that these bytes shipped.
    for event in events:
        if event.sha and event.app and event.kind in ("vercel", "github-deployment", "deploy-job"):
            event.related += [a.id for a in events if a.kind == "artifact" and a.sha == event.sha and a.app == event.app]
    return newest(events)


def candidates(snapshot: dict, events: list[Event], apps: list[str], limit: int, sha: str | None = None):
    commits = {e.sha for e in events if e.sha and e.kind != "release-tag"}
    if sha:
        commits = {sha}
    ordered = sorted(commits, key=lambda s: (max((e.time for e in events if e.sha == s), default=""), s), reverse=True)
    rows = []
    for commit in ordered[:limit]:
        records = [e for e in events if e.sha == commit]
        staging = [e for e in records if e.kind == "deploy-tag" and e.environment == "staging"]
        production = [e for e in records if e.kind == "deploy-tag" and e.environment == "production"]
        gate = snapshot["gates"].get(commit, {"state": "unknown"})
        state = gate["state"]
        classification = "gates-pass" if staging and state == "success" else "blocked" if state in ("failure", "error", "pending") else "insufficient-evidence"
        reasons = (["Verified staging tag"] if staging else ["No verified staging tag in inspected evidence"])
        reasons.append(f"ci/gate-passed: {state}")
        artifacts = {}
        for app in apps:
            builds = [e for e in records if e.kind == "artifact" and e.app == app]
            artifacts[app] = [{"name": e.name, "state": e.state, "url": e.url, "uploaded_at": e.time,
                               "expires_at": e.expires_at, "input_hash": e.input_hash} for e in builds]
        rows.append({"sha": commit, "classification": classification, "reasons": reasons,
                     "staging": [e.time for e in staging], "production": [e.time for e in production],
                     "gate_url": safe_url(gate.get("target_url")), "artifacts": artifacts,
                     "release_tags": [e.name for e in records if e.kind == "release-tag"],
                     "limitations": ["Gates only, not approval or a deployment/health guarantee",
                                     "Same-checkout uploads only; input-hash equivalence and reusable artifacts remain unresolved",
                                     "Production resolves or builds on a miss; landing configuration is environment-specific",
                                     "No manifest/checksum/attestation verification; no Convex compatibility check"]})
    return rows


def event_json(event: Event):
    return asdict(event)

File: scripts/ops/evidence.ts
/** Pure reconciliation. Unknown is preferable to joining on a short SHA or time. */
import { fullSha } from "./types.js";
import type { Candidate, Evidence, Job, Kind, Snapshot } from "./types.js";

export function utc(value: string | number | null | undefined): string {
  if (value === undefined || value === null || value === "") return "";
  const date = new Date(
    typeof value === "string" && !/[Z+-]\d*:?\d*$/.test(value)
      ? `${value}Z`
      : value,
  );
  return Number.isNaN(date.getTime()) ? "" : date.toISOString();
}
export function safeUrl(value: unknown): string {
  if (typeof value !== "string") return "";
  try {
    const parsed = new URL(value);
    if (parsed.protocol !== "https:" || parsed.username || parsed.password)
      return "";
    return `${parsed.origin}${parsed.pathname}`.replace(/\/$/, "");
  } catch {
    return "";
  }
}
export function newest(events: Evidence[]): Evidence[] {
  return [...events].sort(
    (a, b) => b.time.localeCompare(a.time) || b.id.localeCompare(a.id),
  );
}
function event(
  fields: Pick<Evidence, "id" | "kind" | "state" | "source"> &
    Partial<Evidence>,
): Evidence {
  const result = {
    app: null,
    sha: null,
    environment: null,
    time: "",
    url: "",
    updated_at: "",
    context_sha: null,
    name: "",
    input_hash: null,
    expires_at: "",
    current: null,
    notes: [],
    related: [],
    ...fields,
  };
  result.sha ??= null;
  result.environment ??= null;
  return result;
}
export function reconcile(snapshot: Snapshot, apps: string[]): Evidence[] {
  const base = `https://github.com/${snapshot.repo}`;
  const runs = new Map(snapshot.runs.map((r) => [r.id, r]));
  const jobs = new Map(snapshot.jobs.map((j) => [j.id, j]));
  const targets = new Map<number, string | null>();
  const environments = new Map<number, string | null>();
  for (const run of runs.values()) {
    const path = run.path?.split("/").at(-1) ?? "";
    const tags = snapshot.tags.filter(
      (t) => t.run_id === run.id && t.name.startsWith("deploy/"),
    );
    const shas = new Set<string | null>(tags.map((t) => t.sha));
    const envs = new Set(tags.map((t) => t.name.split("/")[1]));
    if (
      (path === "cd-staging.yml" || path.startsWith("ci-")) &&
      ["push", "workflow_dispatch"].includes(run.event ?? "")
    )
      shas.add(fullSha(run.head_sha));
    targets.set(run.id, shas.size === 1 ? [...shas][0] : null);
    environments.set(
      run.id,
      path === "cd-staging.yml"
        ? "staging"
        : path === "cd-production.yml"
          ? "production"
          : envs.size === 1
            ? [...envs][0]
            : null,
    );
  }
  const jobApp = (job?: Job): string | null => {
    const match =
      /^(?:Build|Resolve|Deploy) (.+?)(?: Artifact)? \((?:Staging|Production|Rollback)\)$/.exec(
        job?.name ?? "",
      );
    const app = match?.[1].toLowerCase().replaceAll(" ", "-");
    return app && (apps.includes(app) || app === "convex") ? app : null;
  };
  const events: Evidence[] = [];
  for (const run of runs.values()) {
    const sha = targets.get(run.id) ?? null;
    events.push(
      event({
        id: `run:${run.id}`,
        kind: "workflow",
        sha,
        environment: environments.get(run.id),
        time: utc(run.created_at),
        state: run.conclusion || run.status || "unknown",
        source: run.path?.split("/").at(-1) ?? "",
        url: `${base}/actions/runs/${run.id}`,
        updated_at: utc(run.updated_at),
        context_sha: fullSha(run.head_sha),
        name: `attempt ${run.run_attempt ?? 1}`,
        notes: sha
          ? []
          : [
              "Requested checkout SHA unknown; workflow head is NOT the deployment target",
            ],
      }),
    );
  }
  for (const job of jobs.values()) {
    const run = runs.get(job.run_id);
    const path = run?.path?.split("/").at(-1) ?? "";
    const ci = /^ci-(.+) \/ Build(?: & Bundle Size)?$/.exec(job.name);
    const ciApp =
      ci?.[1] ??
      (path.startsWith("ci-") &&
      ["Build", "Build & Bundle Size"].includes(job.name)
        ? path.slice(3, -4)
        : null);
    if (ciApp && apps.includes(ciApp)) {
      const step = job.steps?.find((s) => s.name === "Build");
      events.push(
        event({
          id: `ci-build:${job.id}`,
          kind: "ci-build",
          app: ciApp,
          sha: targets.get(job.run_id),
          time: utc(step?.started_at || job.started_at),
          state: step?.conclusion || job.conclusion || job.status || "unknown",
          source: `run ${job.run_id}`,
          url: safeUrl(job.html_url),
          updated_at: utc(step?.completed_at || job.completed_at),
          context_sha: fullSha(run?.head_sha),
          name: job.name,
          notes: [
            "CI build only; no deployable bundle implied. PR checkout SHA not inferred from head_sha.",
          ],
        }),
      );
    }
    const app = jobApp(job);
    if (!app) continue;
    events.push(
      event({
        id: `job:${job.id}`,
        kind: job.name.startsWith("Deploy ") ? "deploy-job" : "build-resolve",
        app,
        sha: targets.get(job.run_id),
        environment: environments.get(job.run_id),
        time: utc(job.started_at),
        state: job.conclusion || job.status || "unknown",
        source: `run ${job.run_id}`,
        url: safeUrl(job.html_url),
        updated_at: utc(job.completed_at),
        context_sha: fullSha(run?.head_sha),
        name: job.name,
        notes: ["Job result, not a serving-state or new-build guarantee"],
      }),
    );
  }
  for (const artifact of snapshot.artifacts) {
    const app = [...apps]
      .sort((a, b) => b.length - a.length)
      .find((a) => artifact.name.startsWith(`${a}-`));
    if (!app) continue;
    const match =
      /^(?:(?:staging|production|preview)-)?([0-9a-f]{16}|[0-9a-f]{40})$/.exec(
        artifact.name.slice(app.length + 1),
      );
    if (!match) continue; // Test reports / arbitrary CI output are not prebuilt CD bundles.
    const runId = artifact.workflow_run?.id;
    const expiry = utc(artifact.expires_at);
    const expired =
      artifact.expired === true ||
      (expiry && expiry <= utc(snapshot.gathered_at));
    const sha = runId ? targets.get(runId) : null;
    events.push(
      event({
        id: `artifact:${artifact.id}`,
        kind: "artifact",
        app,
        sha,
        time: utc(artifact.created_at),
        state: expired
          ? "expired"
          : artifact.expired === false && expiry
            ? "available"
            : "unknown",
        source: `build run ${runId ?? "unknown"}`,
        url: runId
          ? `${base}/actions/runs/${runId}/artifacts/${artifact.id}`
          : "",
        context_sha: fullSha(artifact.workflow_run?.head_sha),
        name: artifact.name,
        input_hash: match[1].length === 16 ? match[1] : null,
        expires_at: expiry,
        notes: [
          "Upload time; checkout SHA from workflow evidence, not verified provenance",
          "Manifest, tarball checksum and SLSA attestation not downloaded or verified",
          ...(!sha
            ? ["Builder checkout unknown; artifact head_sha is context only"]
            : []),
        ],
      }),
    );
  }
  for (const tag of snapshot.tags) {
    const deploy = tag.name.startsWith("deploy/");
    events.push(
      event({
        id: `tag:${tag.name}`,
        kind: deploy ? "deploy-tag" : "release-tag",
        sha: tag.sha,
        environment: deploy ? tag.name.split("/")[1] : null,
        time: utc(tag.time),
        state: deploy ? "recorded" : "source-release",
        source: "git tag",
        url: `${base}/tree/${encodeURIComponent(tag.name)}`,
        name: tag.name,
        notes: [
          deploy
            ? "Repository-wide marker; does not prove every app deployed or is healthy"
            : "Starter source version; not a deployable artifact or deployment",
        ],
      }),
    );
  }
  for (const row of snapshot.vercel) {
    const dep = row.deployment;
    const meta = dep.meta ?? {};
    const claims = new Set(
      [fullSha(meta.githubCommitSha), fullSha(meta.gitCommitSha)].filter(
        (s): s is string => s !== null,
      ),
    );
    let sha = claims.size === 1 ? [...claims][0] : null;
    const notes = ["Vercel git metadata; no artifact identity supplied"];
    if (claims.size > 1)
      notes.push("Conflicting commit metadata; SHA withheld");
    if (
      meta.githubCommitOrg &&
      meta.githubCommitRepo &&
      `${meta.githubCommitOrg}/${meta.githubCommitRepo}`.toLowerCase() !==
        snapshot.repo.toLowerCase()
    ) {
      sha = null;
      notes.push("Metadata names a different repository; SHA withheld");
    }
    events.push(
      event({
        id: `vercel:${dep.uid}`,
        kind: "vercel",
        app: row.app,
        sha,
        environment: row.environment,
        time: utc(dep.createdAt ?? dep.created),
        state: dep.readyState || dep.state || "unknown",
        source: `Vercel ${row.project}`,
        url: dep.url
          ? safeUrl(
              dep.url.startsWith("https://") ? dep.url : `https://${dep.url}`,
            )
          : "",
        updated_at: utc(dep.ready ?? dep.readyAt),
        current: row.current,
        notes,
      }),
    );
  }
  for (const dep of snapshot.deployments) {
    const statuses = [...(dep.statuses ?? [])].sort(
      (a, b) =>
        (b.created_at ?? "").localeCompare(a.created_at ?? "") ||
        (b.id ?? 0) - (a.id ?? 0),
    );
    const status = statuses[0];
    const url =
      statuses.map((s) => safeUrl(s.environment_url)).find(Boolean) ?? "";
    const log = statuses.map((s) => safeUrl(s.log_url)).find(Boolean) ?? "";
    const relative = log.startsWith(`${base}/actions/runs/`)
      ? log.slice(base.length)
      : "";
    const match = /^\/actions\/runs\/(\d+)\/job\/(\d+)$/.exec(relative);
    const runId = match ? Number(match[1]) : 0;
    let app = jobApp(match ? jobs.get(Number(match[2])) : undefined);
    let sha = targets.get(runId) ?? null;
    const peers = events.filter(
      (e) =>
        e.kind === "vercel" &&
        url &&
        e.url === url &&
        e.environment === dep.environment &&
        (!app || e.app === app),
    );
    const notes = [
      "GitHub environment status; inactive is not proof of Vercel serving state",
    ];
    if (peers.length === 1) {
      app ||= peers[0].app;
      if (sha && peers[0].sha && sha !== peers[0].sha) {
        notes.push(
          "CONFLICT: workflow target and Vercel commit differ; SHA withheld",
        );
        peers[0].notes.push(
          "CONFLICT: GitHub workflow target differs; inspect both sources",
        );
        sha = null;
      } else sha ||= peers[0].sha;
    }
    if (!sha)
      notes.push(
        "Deployment target unknown; GitHub deployment.sha is context only",
      );
    const evidence = event({
      id: `github-deployment:${dep.id}`,
      kind: "github-deployment",
      app,
      sha,
      environment: dep.environment,
      time: utc(dep.created_at),
      state: status?.state || "unknown",
      source: `GitHub deployment ${dep.id}`,
      url: url || log,
      updated_at: utc(status?.created_at),
      context_sha: fullSha(dep.sha),
      notes,
      related: peers.map((e) => e.id),
    });
    for (const peer of peers) peer.related.push(evidence.id);
    events.push(evidence);
  }
  // Full SHA/app links are browsing aids, NOT claims that these bytes shipped.
  for (const e of events)
    if (
      e.sha &&
      e.app &&
      ["vercel", "github-deployment", "deploy-job"].includes(e.kind)
    ) {
      e.related.push(
        ...events
          .filter(
            (a) => a.kind === "artifact" && a.sha === e.sha && a.app === e.app,
          )
          .map((a) => a.id),
      );
    }
  return newest(events);
}
export function candidates(
  snapshot: Snapshot,
  events: Evidence[],
  apps: string[],
  limit: number,
  sha?: string,
): Candidate[] {
  const commits = sha
    ? [sha]
    : [
        ...new Set(
          events
            .filter((e) => e.sha && e.kind !== "release-tag")
            .map((e) => e.sha!),
        ),
      ];
  const latest = (commit: string): string =>
    events
      .filter((e) => e.sha === commit)
      .map((e) => e.time)
      .sort()
      .at(-1) ?? "";
  commits.sort(
    (a, b) => latest(b).localeCompare(latest(a)) || b.localeCompare(a),
  );
  return commits.slice(0, limit).map((commit) => {
    const records = events.filter((e) => e.sha === commit);
    const staging = records.filter(
      (e) => e.kind === "deploy-tag" && e.environment === "staging",
    );
    const production = records.filter(
      (e) => e.kind === "deploy-tag" && e.environment === "production",
    );
    const gate = snapshot.gates[commit] ?? { state: "unknown" };
    return {
      sha: commit,
      classification:
        staging.length && gate.state === "success"
          ? "gates-pass"
          : ["failure", "error", "pending"].includes(gate.state)
            ? "blocked"
            : "insufficient-evidence",
      reasons: [
        staging.length
          ? "Verified staging tag"
          : "No verified staging tag in inspected evidence",
        `ci/gate-passed: ${gate.state}`,
      ],
      staging: staging.map((e) => e.time),
      production: production.map((e) => e.time),
      gate_url: safeUrl(gate.target_url),
      artifacts: Object.fromEntries(
        apps.map((app) => [
          app,
          records
            .filter((e) => e.kind === "artifact" && e.app === app)
            .map((e) => ({
              name: e.name,
              state: e.state,
              url: e.url,
              uploaded_at: e.time,
              expires_at: e.expires_at,
              input_hash: e.input_hash,
            })),
        ]),
      ),
      release_tags: records
        .filter((e) => e.kind === "release-tag")
        .map((e) => e.name),
      limitations: [
        "Gates only, not approval or a deployment/health guarantee",
        "Same-checkout uploads only; input-hash equivalence and reusable artifacts remain unresolved",
        "Production resolves or builds on a miss; landing configuration is environment-specific",
        "No manifest/checksum/attestation verification; no Convex compatibility check",
      ],
    };
  });
}
export function ofKind(events: Evidence[], kinds: Kind[]): Evidence[] {
  return events.filter((e) => kinds.includes(e.kind));
}
